tests/NLog.UnitTests/Targets/WebServiceTargetTests.cs

```diff
@@ -0,0 +1,244 @@
+            var textBytes = StreamToBytes(textStream);
+
+            textStream.Position = 0;
+            textStream.Flush();
+
+            var resultStream = new MemoryStream();
+            textStream.CopyWithOffset(resultStream, 3);
+            var result = StreamToBytes(resultStream);
+            var expected = textBytes.Skip(3).ToArray();
+            Assert.Equal(result.Length, expected.Length);
+            Assert.Equal(result, expected);
+
+
+        }
+
+        [Fact]
+        public void WebserviceTest_httppost_utf8_default_no_bom()
+        {
+            WebserviceTest_httppost_utf8("", false);
+        }
+
+
+        [Fact]
+        public void WebserviceTest_httppost_utf8_with_bom()
+        {
+            WebserviceTest_httppost_utf8("includeBOM='true'", true);
+        }
+
+        [Fact]
+        public void WebserviceTest_httppost_utf8_no_boml()
+        {
+            WebserviceTest_httppost_utf8("includeBOM='false'", false);
+        }
+
+        private void WebserviceTest_httppost_utf8(string bomAttr, bool includeBom)
+        {
+            var configuration = CreateConfigurationFromString(@"
+                <nlog>
+<targets>
+    <target type='WebService'
+            name='webservice'
+            url='http://localhost:57953/Home/Foo2'
+            protocol='HttpPost'
+          " + bomAttr + @"
+            encoding='UTF-8'
+            methodName='Foo'>
+        <parameter name='empty' type='System.String' layout=''/> <!-- work around so the guid is decoded properly -->
+        <parameter name='guid' type='System.String' layout='${guid}'/>
+        <parameter name='m' type='System.String' layout='${message}'/>
+        <parameter name='date' type='System.String' layout='${longdate}'/>
+        <parameter name='logger' type='System.String' layout='${logger}'/>
+        <parameter name='level' type='System.String' layout='${level}'/>
+        <parameter name='machinename' type='System.String' layout='${machinename}'/>
+    </target>
+</targets>
+                </nlog>");
+
+            var target = configuration.FindTargetByName("webservice") as WebServiceTarget;
+            Assert.NotNull(target);
+
+            Assert.Equal(target.Parameters.Count, 7);
+
+            Assert.Equal(target.Encoding.WebName, "utf-8");
+
+            //async call with mockup stream
+            WebRequest webRequest = WebRequest.Create("http://www.test.com");
+            var request = (HttpWebRequest) webRequest;
+            var streamMock = new StreamMock();
+
+            //event for async testing
+            var counterEvent = new CountdownEvent(1);
+
+            var parameterValues = new object[] {"", "336cec87129942eeabab3d8babceead7", "Debg", "2014-06-26 23:15:14.6348", "TestClient.Program", "Debug", "DELL"};
+            target.DoInvoke(parameterValues, c => counterEvent.Signal(), request,
+                callback =>
+                {
+                    var t = new Task(() => { });
```

Comment:
@ilya-g About the "utf8-nobom"-solution, it wasn't good enough because:

- The unit tests were sometimes wrong, it was testing the preamble (BOM) of the 'Encoding', but it doesn't tell if we are really writing the preamble. 
- The default of the `WebserviceTarget` should be no-BOM. (because of the issues). This is problematic because some users configured `UTF-8` as encoding.
- The `FileTarget` was already writing without BOM, if `utf-8` was configured. I didn't want to change it.